libs-haskell/da-hs-base/src/DA/Signals.hs

```diff
@@ -25,3 +29,15 @@ installSignalHandlers = do
     mainThread <- myThreadId
     void $ installHandler sigTERM (Catch $ throwTo mainThread UserInterrupt) Nothing
 #endif
+
+withCloseOnStdin :: IO a -> IO a
+withCloseOnStdin a = do
+    mainThread <- myThreadId
+    -- We close stdin at the end to avoid potential issues
+    -- with isEOF blocking on Windows.
+    withAsync (go mainThread) (const $ a `finally` hClose stdin)
+  where go mainThread = do
+            b <- isEOF
+            if b
+                then throwTo mainThread UserInterrupt
+                else go mainThread
```

Comment:
Is this a busy loop, or does it not consume cycles constantly?
It might be better to use a short `threadDelay` between checks, or something.

Reply:
`isEOF` blocks until something changes so this is not a busy loop. We will only loop if we get input (unlikely given that we don’t accept legitimate stdin input) .

Reply:
Thanks for the explanation! It might be nice to add a comment.

Reply:
actually that’s not quite true, it will block at the beginning but once we have some input it’s a busy loop. I’ve changed it to have some delay.

Reply:
It looks like it only blocks if there's nothing in the input buffer.

Reply:
exactly

Reply:
Ah, I didn't see your second comment before my last one ^^

Reply:
Another way to prevent the busy loop is to always be reading bytes (`hGetLine`) and handle the `isEOFError` IOError.

Reply:
I considered that but then we are completely screwed if any program invoked by daml-helper tries to read from stdin which seems a bit fragile.